### PR TITLE
Complex-valued coarse preconditioner

### PR DESCRIPTION
**Pull request by simlapointe**
This change uses the true complex-valued system matrix for the coarse solve (sparse direct solve) instead of the real-valued approximation. We still use the real-valued sparse matrix libraries and pass the 2x2 block matrix `[A_r -A_i; A_i A_r]` to the solver. Further speedup could be obtained by using complex-valued sparse matrix libraries, but this would require a bit more work, including updating MFEM interfaces.

### CODE REVIEW

**Review by hughcars — CHANGES_REQUESTED** [review]
Tried it out with MGMaxLevels 1 on the cavity, great that the ksp solves in a single iteration now (with an exact preconditioner effectively). 

I pushed a branch with some suggested changes `hughcars/complex-pc`:
1. A new `SetSubVector` to avoid the use of idx.
2. Reordering some callsites to put IoData ahead of all across the code base. This then means the variadic function avoids using an extra argument for data that was already being pass in as part of `IoData`.

File: palace/linalg/solver.cpp [hunk 1]
```diff
@@ -52,7 +52,34 @@ void MfemWrapperSolver<ComplexOperator>::SetOperator(const ComplexOperator &op)
...
+      block_coeffs(1, 0) = 1.0;
+      block_coeffs(1, 1) = 1.0;
+      A.reset(mfem::HypreParMatrixFromBlocks(blocks, &block_coeffs));
+      idx1.SetSize(op.Width());
+      idx2.SetSize(op.Width());
+      for (int i = 0; i < op.Width(); i++)
+      {
+        idx1[i] = i;
+        idx2[i] = i + op.Width();
+      }
```
[the remark this answers is not in the document]
Comment (simlapointe) [reply palace/linalg/solver.cpp]: Nice!

File: scripts/schema/config/solver.json [hunk 2]
```diff
@@ -117,7 +117,8 @@
         "MGSmoothEigScaleMin": { "type": "number", "minimum": 0 },
         "MGSmoothChebyshev4th": { "type": "boolean" },
         "PCMatReal": { "type": "boolean" },
-        "PCMatShifted": { "type": "boolean" },
+          "PCMatShifted": { "type": "boolean" },
+          "ComplexCoarseSolve": {"type": "boolean"},
```
[the remark this answers is not in the document]
Comment (simlapointe) [reply scripts/schema/config/solver.json]: My bad.

File: palace/linalg/ksp.cpp [hunk 3]
```diff
@@ -152,40 +153,42 @@ ConfigurePreconditionerSolver(MPI_Comm comm, const IoData &iodata,
...
-          iodata, fespaces.GetNumLevels() > 1, fespaces.GetFESpaceAtLevel(0),
-          aux_fespaces->GetFESpaceAtLevel(0), print);
+          complex_coarse, iodata, fespaces.GetNumLevels() > 1,
+          fespaces.GetFESpaceAtLevel(0), aux_fespaces->GetFESpaceAtLevel(0), print);
       break;
     case config::LinearSolverData::Type::BOOMER_AMG:
-      pc = MakeWrapperSolver<OperType, BoomerAmgSolver>(iodata, fespaces.GetNumLevels() > 1,
-                                                        print);
+      pc = MakeWrapperSolver<OperType, BoomerAmgSolver>(complex_coarse, iodata,
+                                                        fespaces.GetNumLevels() > 1, print);
```
[the remark this answers is not in the document]
Comment (simlapointe) [reply palace/linalg/ksp.cpp]: Yes, I like your suggestion.

**Review by hughcars — APPROVED** [review]

**Pull request merged**